core/src/main/kotlin/knote/PageManagerImpl.kt

```diff
@@ -66,12 +60,36 @@ internal class PageManagerImpl(
         }
     }
 
-    override fun evalPage(pageId: String): Pair<PageScript, Any?>? {
+    override fun resultType(pageId: String): KType? {
+        val page = findPage(pageId) ?: run {
+            logger.warn("could not find page $pageId")
+            return null
+        }
+        val pageScript = page.compiledScript ?: run {
+            logger.warn("script for page $pageId is not compiled")
+            evalPage(pageId)?.compiledScript ?: return null
+        }
+        val processFunction =
+            pageScript::class.declaredMemberFunctions.find { it.name == "process" } ?: run {
+                logger.error("no function `process` found in $pageId")
+                return null
+            }
+        return processFunction.returnType
+    }
+
+    override fun findPage(pageId: String): Page? = pages[pageId] ?: run {
+        evalPage(pageId) ?: run {
+            logger.warn("evalPage($pageId) returned null")
+            null
+        }
+    }
+
+    override fun evalPage(pageId: String): Page? {
         val file = notebookScript.fileForPage(pageId, logger) ?: return null
         return evalPage(file, pageId)
     }
 
-    fun evalPage(file: File, id: String = file.name.substringBeforeLast(".page.kts")): Pair<PageScript, Any?>? {
+    fun evalPage(file: File, id: String = file.name.substringBeforeLast(".page.kts")): Page? {
```

Comment:
I'm glad you simplified the return here too

Reply:
this is somewhat still not what you are supposed to call from the GUI, there are better functions for also evaluating the page if necessary

Reply:
oh whoops I can change accordingly, I think I was doing `getResultOrExec`